Add aoti_torch_dtype<> specializations for half and bfloat16

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123692

Fixes #122989. (Note that while the missing symbol issue is fixed, the
test itself is still disabled, because the test runner now segfaults on
`atexit()`; but I think that issue is unrelated to the missing symbol.)

In addition to defining the specializations, I also `= delete`d the
default un-specialized version of `aoti_torch_dtype`, so future missing
dtype references will show up as compile-time instead of link-time
errors.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

int3 committed:
Add aoti_torch_dtype<> specializations for half and bfloat16

Fixes #122989. (Note that while the missing symbol issue is fixed, the
test itself is still disabled, because the test runner now segfaults on
`atexit()`; but I think that issue is unrelated to the missing symbol.)

In addition to defining the specializations, I also `= delete`d the
default un-specialized version of `aoti_torch_dtype`, so future missing
dtype references will show up as compile-time instead of link-time
errors.

[ghstack-poisoned]

int3 commented on test/inductor/test_aot_inductor.py:
```diff
@@ -2577,8 +2577,8 @@ def fail_non_abi_compatible_cuda(is_skip=False):
     "test_simple_dynamic": fail_minimal_arrayref_interface(),
-    # https://github.com/pytorch/pytorch/issues/122989
-    "test_zero_grid_with_unbacked_symbols": fail_with_and_without_stack_allocation(
+    # https://github.com/pytorch/pytorch/issues/123691
+    "test_zero_grid_with_unbacked_symbols": fail_minimal_arrayref_interface(
```
(The non-minimal-arrayref versions of this test were already passing before this PR)

int3 commented:
Filed #123691 to keep track of the segfault

chenyang78 reviewed (approved):
LGTM. Thanks! 

I saw a failing test, which doesn't seem to be related to the change here. Please double-check, @int3 Thanks.

int3 commented:
Looks like another test is passing now, but ultimately the test runner is dying on a segfault right before it exits. Will skip that other test and update the segfault task.

int3 committed:
Update on "Add aoti_torch_dtype<> specializations for half and bfloat16"


Fixes #122989. (Note that while the missing symbol issue is fixed, the
test itself is still disabled, because the test runner now segfaults on
`atexit()`; but I think that issue is unrelated to the missing symbol.)

In addition to defining the specializations, I also `= delete`d the
default un-specialized version of `aoti_torch_dtype`, so future missing
dtype references will show up as compile-time instead of link-time
errors.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

int3 commented:
@pytorchbot merge

int3 commented:
@pytorchbot merge -f "CI machines are not responding"
